converter.go

```diff
@@ -0,0 +1,365 @@
+package converter
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+// ConvertFrom interface allows structs to define custom conversion functions if the automated reflection-based Convert
+// is not able to convert properties due to name changes or other factors.
+type ConvertFrom interface {
+	ConvertFrom(interface{}) error
+}
+
+// Convert takes two objects, e.g. v2_1.Document and &v2_2.Document{} and attempts to map all the properties from one
+// to the other. After the automatic mapping, if a struct implements the ConvertFrom interface, this is called to
+// perform any additional conversion logic necessary.
+func Convert(from interface{}, to interface{}) error {
+	fromValue := reflect.ValueOf(from)
+	fromType := fromValue.Type()
+
+	// handle incoming pointers
+	for fromType.Kind() == reflect.Ptr {
+		fromValue = fromValue.Elem()
+		fromType = fromValue.Type()
+	}
+
+	toValuePtr := reflect.ValueOf(to)
+	toTypePtr := toValuePtr.Type()
+
+	if toTypePtr.Kind() != reflect.Ptr {
+		return fmt.Errorf("to struct provided was not a pointer, unable to set values: %v", to)
+	}
+
+	// toValue must be a pointer, but data needs to be set on the struct directly
+	toValue := toValuePtr.Elem()
+	toType := toValue.Type()
+
+	// Assume structs at this point
```

Comment:
nit: it would be a quick check to see if kind == struct and return an error

Reply:
This got me thinking, and in fact this wasn't necessary at all -- the `Convert` function can even handle primitives. I was able to remove some annoying duplicated code this way, too 👍